drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+	gpio_matrix_in(pin, matrix_in, 0);
+
+	return 0;
+}
+
+static void i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					  u32_t freq_hz)
+{
+	const u32_t period = (APB_CLK_FREQ / freq_hz) / 2;
+
+	*config->regs.scl_low_period = period << I2C_SCL_LOW_PERIOD_S;
+	*config->regs.scl_high_period = period << I2C_SCL_HIGH_PERIOD_S;
+
+	*config->regs.scl_start_h_time =
+		(period / 2) << I2C_SCL_START_HOLD_TIME_S;
+	*config->regs.scl_rstart_s_time =
+		(period / 2) << I2C_SCL_RSTART_SETUP_TIME_S;
+	*config->regs.scl_stop_h_time =
+		(period / 2) << I2C_SCL_STOP_HOLD_TIME_S;
+	*config->regs.scl_stop_s_time =
+		(period / 2) << I2C_SCL_STOP_SETUP_TIME_S;
+
+	*config->regs.sda_hold_time = (period / 4) << I2C_SDA_HOLD_TIME_S;
+	*config->regs.sda_sample_time = (period / 4) << I2C_SDA_SAMPLE_TIME_S;
+}
+
+static int i2c_esp32_configure(struct device *dev, u32_t dev_config_raw)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	union dev_config dev_config = { .raw = dev_config_raw };
+	int key = irq_lock();
+	u32_t v;
+	int r;
+
+	r = i2c_esp32_configure_pins(config->pins.scl,
+				     config->pins.scl_pull_up,
+				     config->pins.default_level,
+				     config->sig.scl_out,
+				     config->sig.scl_in);
+	if (r < 0) {
+		return r;
+	}
+
+	r = i2c_esp32_configure_pins(config->pins.sda,
+				     config->pins.sda_pull_up,
+				     config->pins.default_level,
+				     config->sig.sda_out,
+				     config->sig.sda_in);
+	if (r < 0) {
+		return r;
+	}
+
+	i2c_esp32_set_device_enabled(config, true);
+
+	v = *config->regs.ctr;
+
+	/* MSB or LSB first is configurable for both TX and RX */
+	if (config->mode.tx_lsb_first) {
+		v |= I2C_TX_LSB_FIRST;
+	} else {
+		v &= ~I2C_TX_LSB_FIRST;
+	}
+
+	if (config->mode.rx_lsb_first) {
+		v |= I2C_RX_LSB_FIRST;
+	} else {
+		v &= ~I2C_RX_LSB_FIRST;
+	}
+
+	if (dev_config.bits.is_master_device) {
+		/* Act as master. */
+		v |= I2C_MS_MODE;
+
+		*config->regs.slave_addr = 0;
+	} else {
+		u32_t addr;
+
+		/* Act as slave. */
+		v &= ~I2C_MS_MODE;
+
+		addr = data->address & I2C_SLAVE_ADDR_V;
+		*config->regs.slave_addr |= addr << I2C_SLAVE_ADDR_S;
+		if (dev_config.bits.use_10_bit_addr) {
+			*config->regs.slave_addr |= I2C_ADDR_10BIT_EN;
+		}
+
+		/* Before setting up FIFO and interrupts, stop transmission */
+		*config->regs.ctr &= ~I2C_TRANS_START;
+
+		/* Byte after address isn't the offset address in slave RAM */
+		*config->regs.fifo_cfg &= ~I2C_FIFO_ADDR_CFG_EN;
+	}
+
+	/* Use open-drain for clock and data pins */
+	v |= (I2C_SCL_FORCE_OUT | I2C_SDA_FORCE_OUT);
+
+	v |= I2C_CLK_EN;
+
+	/* Sample data when SCL is high */
+	v &= ~I2C_SAMPLE_SCL_LEVEL;
+
+	*config->regs.ctr = v;
+
+	switch (dev_config.bits.speed) {
+	case I2C_SPEED_STANDARD:
+		i2c_esp32_configure_frequency(config, 100000);
+		break;
+	case I2C_SPEED_FAST:
+		i2c_esp32_configure_frequency(config, 400000);
+		break;
```

Comment:
Contra the extra stuff I flagged in kconfig, this seems to be a spot where you're hard-coding values that should probably be configurable.  Hardware in the wild tends to support wide varieties of speeds, and I'm told people like to tune the bus rates on real devices for things like latency and power.  Probably make these two config variables unless there's some reason the hardware needs exactly 100/400kHz.

Reply:
That's a very good point. By specifying bus speed value as a number we would not only be more flexible but also reduce the code size. This whole case structure could be gone. That said, our current I2C API supports only fixed bus speed values so first we need to change the API. I'll open a Jira issue.